File: org.palladiosimulator.blockchainsystems.core/src/main/java/org/palladiosimulator/blockchainsystems/core/network/P2PLink.java
package org.palladiosimulator.blockchainsystems.core.network;

import org.palladiosimulator.blockchainsystems.core.common.P2PNetworkObject;
import org.palladiosimulator.blockchainsystems.core.common.abstractions.Event;
import org.palladiosimulator.blockchainsystems.core.common.abstractions.ValueProvider;
import org.palladiosimulator.blockchainsystems.core.system.abstractions.Message;

public class P2PLink extends P2PNetworkObject {

    private final ValueProvider<Long> _latencyValueProvider;
    private final ValueProvider<Long> _throughputValueProvider;
    private final P2PNode _firstNode;
    private final P2PNode _secondNode;

    public P2PLink(ValueProvider<Long> latencyValueProvider, ValueProvider<Long> throughputValueProvider, P2PNode firstNode,
                   P2PNode secondNode) {
        _latencyValueProvider = latencyValueProvider;
        _throughputValueProvider = throughputValueProvider;
        _firstNode = firstNode;
        _secondNode = secondNode;
    }

    public void send(P2PNode recipientNode, Message messageContent) {
        if (_firstNode != recipientNode && _secondNode != recipientNode) {
            return;
        }

        P2PNode senderNode = _firstNode == recipientNode ? _secondNode : _firstNode;

        MessageSentEvent msEvent = new MessageSentEvent(
                getSimulationContext().getSystemClock().getCurrentTime(),
                new P2PLinkMessageFrame(
                        messageContent,
                        getSimulationContext().getSystemClock().getCurrentTime()),
                this,
                recipientNode,
                senderNode);

        getSimulationContext()
                .getEventCoordinator()
                .raiseEvent(msEvent);
    }


    public static P2PLink create(ValueProvider<Long> latencyValueProvider, ValueProvider<Long> throughputValueProvider,
                                 P2PNode firstNode, P2PNode secondNode) {
        return new P2PLink(
                latencyValueProvider,
                throughputValueProvider,
                firstNode,
                secondNode);
    }

    @Override
    public void dispatchEvent(Event event) {
        switch (event.getEventType()) {
            case "MessageReceivedEvent":
                handleMessageReceivedEvent((MessageReceivedEvent) event);
                break;
            case "MessageSentEvent":
                handleMessageSentEvent((MessageSentEvent) event);
                break;
        }
    }

    private void handleMessageReceivedEvent(MessageReceivedEvent event) {
        event
                .getRecipientNode()
                .onReceive(
                        event.getMessage().getContent(),
                        event.getSenderNode()
                );
    }

    private void handleMessageSentEvent(MessageSentEvent event) {
        long transmissionDuration = _latencyValueProvider.getValue() +
                (long) event.getMessage().getContent().getSize() / _throughputValueProvider.getValue();

        MessageReceivedEvent newMessageReceivedEvent = new MessageReceivedEvent(
                event.getMessage(),
                getSimulationContext()
                        .getSystemClock()
                        .getCurrentTime() + transmissionDuration,
                this,
                event.getRecipientNode(),
                event.getSenderNode());

        getSimulationContext()
                .getEventCoordinator()
                .raiseEvent(newMessageReceivedEvent);
    }
}

File: org.palladiosimulator.blockchainsystems.core/src/main/kotlin/org/palladiosimulator/blockchainsystems/core/network/P2PLink.kt
package org.palladiosimulator.blockchainsystems.core.network

import org.palladiosimulator.blockchainsystems.core.common.P2PNetworkObject
import org.palladiosimulator.blockchainsystems.core.common.abstractions.Event
import org.palladiosimulator.blockchainsystems.core.common.abstractions.SimulationLifecycleAwareValueProvider
import org.palladiosimulator.blockchainsystems.core.system.abstractions.Message
import kotlin.math.roundToLong

/**
 * Unidirectional link between two [P2PNode]s in a P2P network.
 *
 * @author Davis Riedel, Yannik Sproll
 */
class P2PLink(
  private val latencyValueProvider: SimulationLifecycleAwareValueProvider<Long>,
  private val throughputValueProvider: SimulationLifecycleAwareValueProvider<Long>,
  val fromNode: P2PNode,
  val toNode: P2PNode
) : P2PNetworkObject() {
  override fun onInitialize() {
    super.onInitialize()
    latencyValueProvider.initialize(simulationContext)
    throughputValueProvider.initialize(simulationContext)
  }

  override fun onCleanup() {
    super.onCleanup()
    latencyValueProvider.cleanup()
    throughputValueProvider.cleanup()
  }

  fun send(messageContent: Message) {
    val msEvent = MessageSentEvent(
      simulationContext.systemClock.currentTime,
      P2PLinkMessageFrame(
        messageContent,
        simulationContext.systemClock.currentTime
      ),
      this,
      toNode,
      fromNode
    )

    simulationContext
      .eventCoordinator
      .raiseEvent(msEvent)
  }

  override fun dispatchEvent(event: Event) {
    when (event.eventType) {
      MessageDroppedEvent.EVENT_TYPE -> handleMessageDroppedEvent(event as MessageDroppedEvent)
      MessageReceivedEvent.EVENT_TYPE -> handleMessageReceivedEvent(event as MessageReceivedEvent)
      MessageSentEvent.EVENT_TYPE -> handleMessageSentEvent(event as MessageSentEvent)
    }
  }

  private fun handleMessageReceivedEvent(event: MessageReceivedEvent) {
    event
      .recipientNode
      .onReceive(
        event.message.content,
        event.senderNode
      )
  }

  private fun handleMessageDroppedEvent(event: MessageDroppedEvent) {
    event
      .senderNode
      .onMessageDropped(
        event.message.content,
        event.recipientNode
      )
  }

  private fun handleMessageSentEvent(event: MessageSentEvent) {
    val mbps = throughputValueProvider.getValue() // in Megabits per second (Mbps)

    val event = if (mbps <= 0) {
      // Link failed, raise message dropped event
      MessageDroppedEvent(
        event.message,
        simulationContext.systemClock.currentTime,
        this,
        event.recipientNode,
        event.senderNode
      )
    } else {
      // Link is operational, send message

      val latency = latencyValueProvider.getValue()
      val messageSize = event.message.content.getSize().toLong()

      val bpms = mbps * 1_000_000.0 / 8 // Convert Mbps to bytes per millisecond
      val transmissionDuration = (latency + messageSize / bpms).roundToLong()

      MessageReceivedEvent(
        event.message,
        simulationContext.systemClock.currentTime + transmissionDuration,
        this,
        event.recipientNode,
        event.senderNode
      )
    }

    simulationContext
      .eventCoordinator
      .raiseEvent(event)
  }
}